healthchain/pipeline/basepipeline.py

```diff
@@ -0,0 +1,463 @@
+                Defaults to None.
+            name (str, optional):
+                The name of the component.
+                Defaults to None, in which case the name of the function will be used.
+            input_model (Type[BaseModel], optional):
+                The input Pydantic model class for validating the input data.
+                Defaults to None.
+            output_model (Type[BaseModel], optional):
+                The output Pydantic model class for validating the output data.
+                Defaults to None.
+            dependencies (List[str], optional):
+                The list of component names that this component depends on.
+                Defaults to an empty list.
+
+        Returns:
+            The original component if component is None, otherwise the wrapper function.
+
+        """
+
+        def wrapper(func):
+            def validated_component(data: DataContainer[T]) -> DataContainer[T]:
+                # Validate input if input_model is provided
+                if input_model:
+                    input_model(**data.__dict__)
+
+                # Run the component
+                result = func(data)
+
+                # Validate output if output_model is provided
+                if output_model:
+                    output_model(**result.__dict__)
+
+                return result
+
+            component_func = (
+                validated_component if input_model or output_model else func
+            )
+            new_component = PipelineNode(
+                func=component_func,
+                position=position,
+                reference=reference,
+                stage=stage,
+                name=name
+                if name is not None
+                else (
+                    component_func.__name__
+                    if hasattr(component_func, "__name__")
+                    else (
+                        component_func.__class__.__name__
+                        if hasattr(component_func, "__class__")
+                        else str(component_func)
+                    )
+                ),
+                dependencies=dependencies,
+            )
+            try:
+                self._add_component_at_position(new_component, position, reference)
+            except Exception as e:
+                raise ValueError(f"Error adding component: {str(e)}")
+
+            if stage:
+                if stage not in self._stages:
+                    self._stages[stage] = []
+                self._stages[stage].append(func)
+                logger.debug(
+                    f"Successfully added component '{new_component.name}' to stage '{stage}'."
+                )
+
+            return func
+
+        if component is None:
+            return wrapper
+        if isinstance(component, BaseComponent):
+            return wrapper(component)
+        elif isinstance(component, Component):
+            return wrapper(component.__call__)
```

Comment:
I don't understand why component needs __call__ if it's not a basecomponent.

Reply:
good point